src/passes/Heap2Local.cpp

```diff
@@ -0,0 +1,617 @@
+      // falls through later to check for mixing.
+      //
+      // Note that this does not need to be set for expressions if their type
+      // proves that the value does not continue onwards (e.g. if their type is
+      // none, or not a reference type), but for clarity some do still mark this
+      // field as true when it is clearly so.
+      bool fullyConsumes = false;
+
+      // General operations
+      void visitBlock(Block* curr) {
+        escapes = false;
+        // We do not mark fullyConsumes as the value may continue through this
+        // and other control flow structures.
+      }
+      // Note that If is not supported here, because for our value to flow
+      // through it there must be an if-else, and that means there is no single
+      // value falling through anyhow.
+      void visitLoop(Loop* curr) { escapes = false; }
+      void visitDrop(Drop* curr) {
+        escapes = false;
+        fullyConsumes = true;
+      }
+      void visitBreak(Break* curr) { escapes = false; }
+      void visitSwitch(Switch* curr) { escapes = false; }
+
+      // Local operations. Locals by themselves do not escape; the analysis
+      // tracks where locals are used.
+      void visitLocalGet(LocalGet* curr) { escapes = false; }
+      void visitLocalSet(LocalSet* curr) { escapes = false; }
+
+      // Reference operations. TODO add more
+      void visitRefAs(RefAs* curr) {
+        // TODO General OptimizeInstructions integration, that is, since we know
+        //      that our allocation is what flows into this RefAs, we can
+        //      know the exact outcome of the operation.
+        if (curr->op == RefAsNonNull) {
+          // As it is our allocation that flows through here, we know it is not
+          // null, and can mark it as safe to optimize.
+          //
+          // Note that while we can look through it, we cannot optimize it out
+          // later: the outside might depend on receiving a non-nullable type
+          // (even though we have proven that it cannot be null).
+          escapes = false;
+        }
+      }
+
+      // GC operations.
+      void visitStructSet(StructSet* curr) {
+        // The reference does not escape (but the value is stored to memory and
+        // therefore might).
+        if (curr->ref == child) {
+          escapes = false;
+          fullyConsumes = true;
+        }
+      }
+      void visitStructGet(StructGet* curr) {
+        escapes = false;
+        fullyConsumes = true;
+      }
+
+      // TODO Array and I31 operations
+    } checker;
+
+    checker.child = child;
+    checker.visit(parent);
+
+    if (checker.escapes) {
+      return ParentChildInteraction::Escapes;
+    }
+
+    // If the parent returns a type that is not a reference, then by definition
+    // it fully consumes the value as it does not flow our allocation onward.
+    if (checker.fullyConsumes || !parent->type.isRef()) {
```

Comment:
Relying on `fullyConsumed` is an option. But the case of not returning a reference at all is easy to handle, and saves writing more code for all the classes.

I'm not sure why a supertype sounds right to you. The parent could also flow a subtype (as `ref.as_non_null` etc. does). As long as it's a reference, it could be our reference in theory, that's what this line is saying basically.

About tuples, they are ruled out by not appearing in the code above (since we assume `escapes` unless overridden). To support tuples we'd need to add code there (and then not use this line, correct).